File: src/contest/ccc/CCC_2015_Stage_2_Solar_Flight.cpp
#include <bits/stdc++.h>

using namespace std;

#define SIZE 2001
#define EPS 0.000001
typedef long long ll;

struct query {
    int index, plane, position;
    query (int index_, int plane_, int position_) {
        index = index_;
        plane = plane_;
        position = position_;
    }
};

struct collision {
    double startPos, endPos;
    int inc, dec;
    ll val;
    collision (double pos, ll val_) {
        startPos = pos;
        val = val_;
    }
    collision (double pos, int inc_, int dec_) {
        startPos = pos;
        inc = inc_;
        dec = dec_;
    }
};

class cmp_query {
    public:
        bool operator() (const query& q1, const query& q2) const {
            return q1.position < q2.position;
        }
};
class cmp_collision {
    public:
        bool operator() (const collision& c1, const collision& c2) const {
            return c1.startPos > c2.startPos;
        }
};

int x, k, n, q;
int a[SIZE], b[SIZE];
ll c[SIZE];
vector<deque<collision>> maxState (SIZE);
ll curr[SIZE];
priority_queue<collision, vector<collision>, cmp_collision> pq;

void add (collision c, int plane) {
    while (!maxState[plane].empty() && (maxState[plane].back().val <= c.val || abs(maxState[plane].back().startPos - c.startPos) <= EPS))
        maxState[plane].pop_back();
    maxState[plane].push_back(c);
}

int main () {
    scanf("%d%d%d%d", &x, &k, &n, &q);
    for (int i = 0; i < n; i++) {
        maxState[i].push_back(collision(0, 0));
        scanf("%d%d%lld", &a[i], &b[i], &c[i]);
    }
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (a[i] > a[j])
                pq.push(collision(0.0, j, i));
            else
                pq.push(collision(0.0, i, j));
            double s1 = (double)(b[i] - a[i])/x;
            double s2 = (double)(b[j] - a[j])/x;
            double cx = (double)(a[j] - a[i])/(s1 - s2);
            if (0 <= cx && cx <= x) {
                if (b[i] > b[j])
                    pq.push(collision(cx, j, i));
                else
                    pq.push(collision(cx, i, j));
            }
        }
    }
    vector<query> queries;
    for (int i = 0; i < q; i++) {
        int plane, pos;
        scanf("%d%d", &plane, &pos);
        queries.push_back(query(i, plane-1, pos));
    }
    sort(queries.begin(), queries.end(), cmp_query());
    ll ans[q];
    for (query qu : queries) {
        while (!pq.empty() && pq.top().startPos <= qu.position + k) {
            collision next = pq.top();
            pq.pop();
            curr[next.inc] += c[next.dec];
            maxState[next.inc].back().endPos = next.startPos;
            add(collision(next.startPos, curr[next.inc]), next.inc);
            if (next.startPos != 0) {
                curr[next.dec] -= c[next.inc];
                maxState[next.dec].back().endPos = next.startPos;
                add(collision(next.startPos, curr[next.dec]), next.dec);
            }
        }
        //for (collision co : maxState[qu.plane]) {
        //    printf("(%f, %d) ", co.startPos, co.val);
        //}
        while (maxState[qu.plane].size() > 1 && maxState[qu.plane][0].endPos <= qu.position)
            maxState[qu.plane].pop_front();
        ans[qu.index] = maxState[qu.plane].front().val;
    }
    for (int i = 0; i < q; i++)
        printf("%lld\n", ans[i]);
    return 0;
}

File: src/contest/ccc/CCC_2015_Stage_2_Solar_Flight.java
package contest.ccc;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class CCC_2015_Stage_2_Solar_Flight {

  static final double EPS = 0.000001;
  static BufferedReader br;
  static PrintWriter out;
  static StringTokenizer st;
  static int x, k, n, q;
  static int[] a, b, c;
  static ArrayList<LinkedList<Collision>> max = new ArrayList<LinkedList<Collision>>();
  static PriorityQueue<Collision> col = new PriorityQueue<Collision>();
  static long[] curr;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    out = new PrintWriter(new OutputStreamWriter(System.out));
    //br = new BufferedReader(new FileReader("in.txt"));
    //out = new PrintWriter(new FileWriter("out.txt"));

    x = readInt();
    k = readInt();
    n = readInt();
    q = readInt();
    a = new int[n];
    b = new int[n];
    c = new int[n];
    curr = new long[n];
    for (int i = 0; i < n; i++) {
      max.add(new LinkedList<Collision>());
      max.get(i).add(new Collision(0, 0));
      a[i] = readInt();
      b[i] = readInt();
      c[i] = readInt();
    }

    for (int i = 0; i < n; i++) {
      for (int j = i + 1; j < n; j++) {
        if (a[i] > a[j]) {
          col.add(new Collision(0.0, j, i));
        } else if (a[j] > a[i]) {
          col.add(new Collision(0.0, i, j));
        }
        double s1 = (double)(b[i] - a[i]) / x;
        double s2 = (double)(b[j] - a[j]) / x;
        double cx = (double)(a[j] - a[i]) / (s1 - s2);
        if (0 <= cx && cx <= x) {
          if (b[i] > b[j])
            col.offer(new Collision(cx, j, i));
          else
            col.offer(new Collision(cx, i, j));
        }
      }
    }
    ArrayList<Query> queries = new ArrayList<Query>();
    for (int i = 0; i < q; i++)
      queries.add(new Query(i, readInt() - 1, readInt()));
    Collections.sort(queries);
    long[] ans = new long[q];
    for (Query query : queries) {
      while (!col.isEmpty() && col.peek().pos <= query.position + k) {
        Collision next = col.poll();
        curr[next.inc] += c[next.dec];
        max.get(next.inc).getLast().end = next.pos;
        add(new Collision(next.pos, curr[next.inc]), next.inc);
        if (next.pos != 0) {
          curr[next.dec] -= c[next.inc];
          max.get(next.dec).getLast().end = next.pos;
          add(new Collision(next.pos, curr[next.dec]), next.dec);
        }
      }
      while (max.get(query.plane).size() > 1 && max.get(query.plane).get(0).end <= query.position)
        max.get(query.plane).pollFirst();
      ans[query.index] = max.get(query.plane).getFirst().val;
    }
    for (int i = 0; i < q; i++)
      out.println(ans[i]);

    out.close();
  }

  static void add(Collision c, int plane) {
    while (!max.get(plane).isEmpty() && (max.get(plane).getLast().val <= c.val || Math.abs(max.get(plane).getLast().pos - c.pos) < EPS))
      max.get(plane).pollLast();
    max.get(plane).addLast(c);
  }

  static String next() throws IOException {
    while (st == null || !st.hasMoreTokens())
      st = new StringTokenizer(br.readLine().trim());
    return st.nextToken();
  }

  static long readLong() throws IOException {
    return Long.parseLong(next());
  }

  static int readInt() throws IOException {
    return Integer.parseInt(next());
  }

  static double readDouble() throws IOException {
    return Double.parseDouble(next());
  }

  static char readCharacter() throws IOException {
    return next().charAt(0);
  }

  static String readLine() throws IOException {
    return br.readLine().trim();
  }

  static class Query implements Comparable<Query> {
    int index, plane, position;

    Query(int index, int plane, int position) {
      this.index = index;
      this.plane = plane;
      this.position = position;
    }

    @Override
    public int compareTo(Query q) {
      return position - q.position;
    }
  }

  static class Collision implements Comparable<Collision> {
    Double pos, end;
    int inc, dec;
    long val;

    Collision(double pos, long val) {
      this.pos = pos;
      this.val = val;
    }

    Collision(double pos, int inc, int dec) {
      this.pos = pos;
      this.inc = inc;
      this.dec = dec;
    }

    @Override
    public int compareTo(Collision c) {
      return pos.compareTo(c.pos);
    }
  }
}
